Sources/ClientRuntime/Client/ClientBuilder.swift

```diff
@@ -4,29 +4,36 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 //
+
 public class ClientBuilder<ClientType: Client> {
 
-    private var plugins: [Plugin]
+    private struct PluginContainer: Plugin {
+        let plugin: any Plugin<ClientType.Config>
 
-    public init(defaultPlugins: [Plugin] = []) {
-        self.plugins = defaultPlugins
+        func configureClient(clientConfiguration: inout ClientType.Config) async throws {
+            try await plugin.configureClient(clientConfiguration: &clientConfiguration)
+        }
     }
 
-    public func withPlugin(_ plugin: any Plugin) -> ClientBuilder<ClientType> {
-        self.plugins.append(plugin)
+    private var plugins = [PluginContainer]()
+
+    public init() {}
+
+    public func withPlugin<P: Plugin>(_ plugin: P) -> ClientBuilder<ClientType> where P.Config == ClientType.Config {
+        self.plugins.append(PluginContainer(plugin: plugin))
         return self
     }
```

Comment:
Plugins are added with a `.withPlugin()` method so that each can be placed in a type-erased container for storage in an array.